### ISSUE TITLE
Models input/output conventions (shape, preproc/postproc, etc.)

### ISSUE THREAD

**Issue by NicolasHug** [pytorch/vision#8126]
Not an issue, just writing this down for reference.

(**TODO** perhaps it'd be useful to clarify whether models support arbitrary H and W)

## Img Classification

**Input**
Shape (N, C, H, W).
Must be scaled from [0, 1] into specific scale

**Output**
Shape (N, num_classes), logits

**preproc and postproc within model** None

## Semantic segmentation

**Input**
Shape (N, C, H, W). The expected H,W as per weight preproc is (520, 520), not enforced.
Must be scaled from [0, 1] into specific scale

**Output** (Both training and eval)
OrderedDict with keys:
- "out", shape (N, num_classes, H, W) logits where H, W are the input image
  size.
- "aux", only for deeplabv3 and fcn and if aux_loss is True.  Same shape as "out".

**preproc  within model** None

**postproc within model** The masks are predicted at lower resolutions so they get  upsampled at the end of `forward()`. Note that the `weights.transforms()` may resize the input, which is the shape that the model will then output, and may be different from the original image size.

## Video classification 

**Input**
Shape (N, C, T, H, W). T = number of frames in each clip / sample. C is num_channels as usual.
Must be scaled from [0, 1] into custom scale

**Output**
(N, num_classes) logits. 

## Optical Flow

**Input**
2 imgs of shape (N, C, H, W). H, W must be multiples of 8.
Scale: [-1, 1]
Note that the weight transforms don't resize so they don't ensure the input is
divisible by 8.

**Output**
list of tensors of shape (N, 2, H, W). 2 channels for horiz and vertical directions. len(list) = num_flow_updates.

**postproc within model** predicted flows are up-sampled to the original image size. This is just a call to `interpolate()` for raft_small but for raft_large, the upsampling is *learned* during training so it needs to be done in the model.

## Object detection

This includes object detection (Mask-RCNN) and ignores keypoint detection which is fairly similar.

**Input**
list of images of shape (C, H, W). Each image can have a different shape (it will be pre-processed). len(list) == N.
scale: [0, 1]
During training, must also pass target as a list (len=N) of dicts where each dict has keys:
- "boxes": shape (num_instances, 4) in XYXY format
- "labels": shape (num_instances,)
- "masks": shape (num_instances, H, W) for instance seg.

**Output**
- when training: dict[str, Tensor] of losses, e.g. for fasterrcnn:
 - "loss_classifier"
 - "loss_box_reg"
 - "loss_objectness"
 - "loss_rpn_box_reg"
 - "loss_mask" for instance seg
- when eval: list (len=N) of dicts with keys:
  - "boxes": shape (num_det, 4) in XYXY format
  - "labels": shape (num_det,)
  - "scores": shape (num_det,)
  - "masks": shape (num_det, 1, H, W) for instance seg. Value aren't logits, they're probabilities of each pixel being part of the instance.

Since forever, users have requested the output to be unified https://github.com/pytorch/vision/issues/1574 across training and eval.

These are the only models where the loss is computed within the model, and not outside of it.

**preproc within model** (training and eval)
- scaling from [0, 1] to model-dependent scale
- resizing of image (for kp-rcnn, randomly resizing for augmentation) based on
  model-depedent min_size and max_size. Even after resizing, images may still
  differ in size. They are copied into a NCHW batch where HW depends on the max
  dims of each image *and* the `size_divisible` param. masks and boxes are
  resized accordingly. ImageList contains that batch along with the sizes of the
  images *after* they were resized: this is what the model will use internaly,
  and those sizes will be needed as well in postproc.

**postproc within model**
  - during training, nothing is done. There's no need, these models only return
    the losses and they can be computed on the pre-processed inputs.
  - during eval: boxes and masks are resized back to the original image size.
